github_pr_to_internal_pr/github_pr_to_internal_pr.py

```diff
@@ -0,0 +1,237 @@
+    new_cmt_msg = commit.message + '\nMerges ' + pr_html_url
+
+    print('Amending commit message (Adding additional info about commit)...')
+    print(git.execute(['git','commit', '--amend', '-m', new_cmt_msg]))
+
+    print('Pushing to remote...')
+    print(git.push('--set-upstream', 'origin', pr_branch))
+
+
+def main():
+    if 'GITHUB_REPOSITORY' not in os.environ:
+        print('Not running in GitHub action context, nothing to do')
+        return
+
+    if not os.environ['GITHUB_REPOSITORY'].startswith('espressif/'):
+        print('Not an Espressif repo!')
+        return
+
+    # The path of the file with the complete webhook event payload. For example, /github/workflow/event.json.
+    with open(os.environ['GITHUB_EVENT_PATH'], 'r') as f:
+        event = json.load(f)
+        # print(json.dumps(event, indent=4))
+
+    event_name = os.environ['GITHUB_EVENT_NAME']  # The name of the webhook event that triggered the workflow.
+    action = event["action"]
+    state = event["review"]["state"]
+    review_body = event["review"]["body"]
+
+    if event_name != 'pull_request_review' or state != 'approved':
+        raise SystemError("False Trigger!")
+
+    pr_base = event["pull_request"]["base"]["ref"]
+    if pr_base != 'master':
+        raise SystemError("PR base illegal! Should be the master branch!")
+    
+    project_fullname = event["repository"]["full_name"]
+    project_org, project_name = project_fullname.split("/")
+    project_users_url = event["repository"]["collaborators_url"]
+    project_html_url = event["repository"]["clone_url"]
+
+    pr_num = event["pull_request"]["number"]
+    pr_branch = 'contrib/github_pr_' + str(pr_num)
+    pr_rest_url = event["pull_request"]["url"]
+    pr_html_url = event["pull_request"]["html_url"]
+
+    pr_files_url = pr_rest_url + '/files'
+    # Check whether the PR has modified forbidden files
+    pr_check_forbidden_files(pr_files_url)
+
+    pr_approver = event["review"]["user"]["login"]
+    # Checks whether the approver access level is above required; needs Github access token
+    pr_check_approver_access(project_users_url, pr_approver)
+
+    # Getting the PR title and body
+    pr_title = event["pull_request"]["title"]
+    idx = pr_title.find(os.environ['JIRA_PROJECT']) # Finding the JIRA issue tag
+    pr_title_desc = pr_title[0 : idx - 2] # For space character
+    pr_jira_issue = pr_title[idx : -1]
+    pr_body = event["pull_request"]["body"]
+
+    # Gitlab setup and cloning internal codebase
+    gl = setup_project(project_fullname)
+
+    if "/rebase" in review_body:
+        sync_pr_with_rebase(project_name, pr_branch, pr_html_url, pr_rest_url)
+    elif "/merge" in review_body:
+        sync_pr_with_merge(project_name, pr_num, pr_branch, project_html_url)
+    else:
+        print('No action selected!!!')
+        return
+
+    # Deleting local repo
+    shutil.rmtree(project_name)
+
+    # # NOTE: Remote takes some time to register a branch
```

Comment:
... then maybe we could check in a loop:

```python
for _ in range(15):
    # exists?
    time.sleep(1)
else:
    # raise
```

Reply:
Hello @dobairoland. I tried using this approach ([here](https://github.com/laukik-hase/github-actions/blob/a355b1b3468f72ebd22fcaab477a22f6f57a8af0/github_pr_to_internal_pr/github_pr_to_internal_pr.py#L72)), but it seems there is a bug in GitLab. After pushing the branch, the remote takes time to register the branch. But the API (which says whether a branch exists or not) gives out True immediately. This leads to creation of an empty MR. When the branch is registered finally, multiple pipelines for the same commit start running. You can take a look at these issues on Gitlab: [17381](https://gitlab.com/gitlab-org/gitlab/-/issues/17381) and [35752](https://gitlab.com/gitlab-org/gitlab/-/issues/35752).

Reply:
Could you please take a look at the internal backporting script? Local branch is created, pushed and no delay is necessary before creating the MR. The Repo object is used instead of Git. The Git object seem to be derived from a "lazy" class.

Reply:
Hello, @dobairoland. I checked the internal backporting script and it also shows the same behavior as I have mentioned. I will try to find to more about this issue; as I agree that the current fix is inelegant.

Reply:
Hello, @dobairoland. I tried to find out more about this, but could not get it this issue to go away. Can we for the time being accept this inelegant fix (as this is a small issue) and test out rest of the PR feature? As soon as we get results, we can try to fix all of tthese issues. What do you suggest?

Reply:
Yes, we can. But I'd like to suggest to extend the comment by reminding us what has been tried and didn't work. (In order to save time for anyone who will try to fix this)

Reply:
Yes, I agree. So, should we proceed with this feature?

PS: I have updated the README with a To-Do list, PTAL ([here](https://github.com/espressif/github-actions/pull/17/files#diff-9574bf2c0698b4d6eaacccf04a14299f665c674f798a2d6f9c678b2871e88ae3R12)).